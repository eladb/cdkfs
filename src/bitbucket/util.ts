

export function reduceRenderObject( obj:Record<string, any> ) {

  const result = Object.keys(obj).reduce( ( inp:Record<string, any>, key:string) => {

    if (obj[ key ] !== undefined) inp[ key ] = obj[ key ];

    return inp;
  }, {} );

  if ( Object.keys( result).length === 0 ) return undefined;

  return result;
}


export function reduceRenderArray( arr:any[]|undefined ) {

  if ( arr === undefined ) return arr;
  else if (arr.length === 0) return undefined;
  return arr;
}