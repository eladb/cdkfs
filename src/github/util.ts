export function secretToString(secretName: string): string {
  return `\${{ secrets.${secretName} }}`;
}

/**
 * Bash if equivalent execution of scripts that is compatible with shx.
 * Execution will fail, if the when or then command fails.
 *
 * @param when The condition to evaluate. Usually a `test` command. This command may be executed twice. Make sure it is non mutating.
 * @param then The script to run if the condition is true.
 * @param otherwise The script to run if the condition is false.
 */
export function shxIf(when: string, then: string, otherwise?: string): string {
  // Connect the condition and the 'then' command with a logical AND.
  // If the condition is true, this will execute the 'then' command.
  // This case is equivalent to TRUE && then()
  // If the condition is false, this will exit with a non-zero status code
  // (indicating an error) without executing the 'then' command.
  // This case is equivalent to FALSE && then()
  const if_clause = `${when} && (${then})`;

  // For the else clause, we need to check the 'when' condition again,
  // because the if clause can also return FALSE if the 'then' command fails.
  // However in this situation we don't want to execute the else clause but report an error.
  const else_clause = `${when} || (${otherwise})`;

  // Try the if_clause first. If it fails try the else clause.
  if (otherwise) {
    return `(${if_clause}) || (${else_clause})`;
  }

  // Just run the if clause
  return if_clause;
}
