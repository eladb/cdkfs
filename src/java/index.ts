export * from './pom';
export * from './maven-project';
export * from './maven-sample';
export * from './maven-compile';
export * from './maven-versions';
export * from './maven-jar';
export * from './junit';