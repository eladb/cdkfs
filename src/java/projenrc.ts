import { dirname, join } from 'path';
import { existsSync, mkdirpSync, readJsonSync, writeFileSync } from 'fs-extra';
import { PROJEN_VERSION } from '../common';
import { Component } from '../component';
import { DependencyType } from '../deps';
import { Project } from '../project';
import { Pom } from './pom';

/**
 * Options for `Projenrc`.
 */
export interface ProjenrcCommonOptions {
  /**
   * The name of the Java class which contains the `main()` method for projen.
   * @default "projenrc"
   */
  readonly className?: string;

  /**
   * The projen version to use
   * @default - current version
   */
  readonly projenVersion?: string;

  /**
   * Defines projenrc under the test scope instead of the main scope, which is
   * reserved to the app. This means that projenrc will be under
   * `src/test/java/projenrc.java` and projen will be defined as a test
   * dependency. This enforces that application code does not take a dependency
   * on projen code.
   *
   * If this is disabled, projenrc should be under
   * `src/main/java/projenrc.java`.
   *
   * @default true
   */
  readonly testScope?: boolean;
}

export interface ProjenrcOptions extends ProjenrcCommonOptions {
  /**
   * Project initialization options.
   */
  readonly initializationOptions?: { [name: string]: any };
}

/**
 * Allows writing projenrc files in java.
 *
 * This will install `org.projen/projen` as a Maven dependency and will add a
 * `synth` task which will compile & execute `main()` from
 * `src/main/java/projenrc.java`.
 */
export class Projenrc extends Component {
  /**
   * The name of the java class that includes the projen entrypoint.
   */
  public readonly className: string;

  /**
   * Whether the class is in the "test" or "main" scope?
   */
  private readonly testScope: boolean;

  constructor(project: Project, pom: Pom, options: ProjenrcOptions = {}) {
    super(project);

    const projenVersion = options.projenVersion ?? PROJEN_VERSION;
    this.className = options.className ?? 'projenrc';
    this.testScope = options.testScope ?? true;

    const depType = this.testScope ? DependencyType.TEST : DependencyType.RUNTIME;
    const execOpts = this.testScope ? ' -Dexec.classpathScope="test"' : '';
    const compileGoal = this.testScope ? 'compiler:testCompile' : 'compiler:compile';

    project.deps.addDependency(`com.github.eladb/projen@${projenVersion}`, depType);
    pom.addPlugin('org.codehaus.mojo/exec-maven-plugin@3.0.0');

    // set up the "default" task which is the task executed when `projen` is executed for this project.
    const defaultTask = project.addTask(Project.DEFAULT_TASK, { description: 'Synthesize the project' });
    defaultTask.exec(`mvn ${compileGoal} --quiet`);
    defaultTask.exec(`mvn exec:java --quiet -Dexec.mainClass=${this.className}${execOpts}`);

    // if this is a new project, generate a skelaton for projenrc.java
    this.generateProjenrc(options.initializationOptions);
  }

  private generateProjenrc(initOptions?: Record<string, any>) {
    if (!this.project.jsiiFqn) {
      return; // cannot generate projenrc without the FQN of the project type.
    }

    let [moduleName] = this.project.jsiiFqn.split('.');
    if (moduleName === 'projen') {
      moduleName = '../..';
    }

    const jsiiManifestFile = require.resolve(`${moduleName}/.jsii`);
    const jsiiManifest = readJsonSync(jsiiManifestFile);
    const jsiiType = jsiiManifest.types[this.project.jsiiFqn];
    const javaTarget = jsiiManifest.targets.java;
    const optionsTypeFqn = jsiiType.initializer?.parameters?.[0].type?.fqn;
    if (!optionsTypeFqn) {
      this.project.logger.warn('cannot determine jsii type for project options');
      return;
    }
    const jsiiOptionsType = jsiiManifest.types[optionsTypeFqn];
    if (!jsiiOptionsType) {
      this.project.logger.warn(`cannot find jsii type for project options: ${optionsTypeFqn}`);
      return;
    }

    const dir = this.testScope ? 'src/test/java' : 'src/main/java';
    const split = this.className.split('.');
    let javaClass: string, javaPackage: string[];
    if (split.length === 1) {
      javaClass = split[0];
      javaPackage = [];
    } else {
      javaPackage = split.slice(0, split.length - 2);
      javaClass = split[split.length - 1];
    }

    const javaFile = join(this.project.outdir, dir, ...javaPackage, javaClass + '.java');

    // skip if file exists
    if (existsSync(javaFile)) {
      return;
    }

    const lines = new Array<string>();

    let indent = 0;
    const emit = (line: string = '') => lines.push(' '.repeat(indent * 4) + line);
    const openBlock = (line: string = '') => { emit(line + ' {'); indent++; };
    const closeBlock = () => { indent--; emit('}'); };

    if (javaPackage.length > 0) {
      emit(`package ${javaPackage.join('.')};`);
      emit();
    }

    emit(`import ${javaTarget.package}.${toJavaFullTypeName(jsiiType)};`);
    emit(`import ${javaTarget.package}.${toJavaFullTypeName(jsiiOptionsType)};`);
    emit();
    openBlock(`public class ${javaClass}`);
    openBlock('public static void main(String[] args)');
    emit(`${jsiiType.name} project = new ${jsiiType.name}(${renderJavaOptions(indent, jsiiOptionsType.name, initOptions)});`);
    emit('project.synth();');
    closeBlock();
    closeBlock();

    mkdirpSync(dirname(javaFile));
    writeFileSync(javaFile, lines.join('\n'));
  }
}

function renderJavaOptions(indent: number, optionsTypeName: string, initOptions?: Record<string, any>): string {
  if (!initOptions || Object.keys(initOptions).length === 0) {
    return ''; // no options
  }

  const lines = [`${optionsTypeName}.builder()`];

  for (const [name, value] of Object.entries(initOptions)) {
    lines.push(`.${toJavaProperty(name)}(${toJavaValue(value)})`);
  }

  lines.push('.build()');

  return lines.join(`\n${' '.repeat((indent + 1) * 4)}`);
}

function toJavaProperty(prop: string) {
  return prop;
}

function toJavaValue(value: any) {
  return JSON.stringify(value);
}

function toJavaFullTypeName(jsiiType: any) {
  return [jsiiType.namespace, jsiiType.name].filter(x => x).join('.');
}