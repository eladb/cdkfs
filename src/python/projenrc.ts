import { dirname, join } from 'path';
import { existsSync, mkdirpSync, writeFileSync } from 'fs-extra';
import { PROJEN_VERSION } from '../common';
import { Component } from '../component';
import { DependencyType } from '../deps';
import { readJsiiManifest } from '../inventory';
import { Project } from '../project';

// eslint-disable-next-line @typescript-eslint/no-require-imports
const decamelize = require('decamelize');

/**
 * Options for `Projenrc`.
 */
export interface ProjenrcOptions {
  /**
   * The name of the projenrc file.
   * @default ".projenrc.py"
   */
  readonly filename?: string;

  /**
   * The projen version to use
   * @default - current version
   */
  readonly projenVersion?: string;
}

/**
 * Allows writing projenrc files in python.
 *
 * This will install `projen` as a Python dependency and will add a
 * `synth` task which will run `.projenrc.py`.
 */
export class Projenrc extends Component {
  /**
   * The name of the projenrc file.
   */
  private readonly rcfile: string;

  constructor(project: Project, options: ProjenrcOptions = {}) {
    super(project);

    const projenVersion = options.projenVersion ?? PROJEN_VERSION; // ?
    this.rcfile = options.filename ?? '.projenrc.py';

    project.deps.addDependency(`projen@${projenVersion}`, DependencyType.DEVENV);

    // set up the "default" task which is the task executed when `projen` is executed for this project.
    const defaultTask = project.addTask(Project.DEFAULT_TASK, { description: 'Synthesize the project' });
    defaultTask.exec('python projenrc.py');

    // if this is a new project, generate a skeleton for projenrc.py
    this.generateProjenrc();
  }

  private generateProjenrc() {
    const bootstrap = this.project.newProject;
    if (!bootstrap) {
      return;
    }
    const jsiiFqn = bootstrap.fqn;
    const jsiiManifest = readJsiiManifest(jsiiFqn);
    const jsiiType = jsiiManifest.types[jsiiFqn];
    const pythonTarget = jsiiManifest.targets.python;
    const optionsTypeFqn = jsiiType.initializer?.parameters?.[0].type?.fqn;
    if (!optionsTypeFqn) {
      this.project.logger.warn('cannot determine jsii type for project options');
      return;
    }

    const pythonFile = join(this.project.outdir, this.rcfile);

    // skip if file exists
    if (existsSync(pythonFile)) {
      return;
    }

    const lines = new Array<string>();

    let indent = 0;
    const emit = (line: string = '') => lines.push(' '.repeat(indent * 4) + line);
    // const openBlock = (line: string = '') => { emit(line + ' {'); indent++; };
    // const closeBlock = () => { indent--; emit('}'); };

    emit(toPythonImport(pythonTarget.module, jsiiType));
    emit();
    emit(`project = ${jsiiType.name}(${renderPythonOptions(indent, bootstrap.args)});`);
    emit();
    emit('project.synth();');

    mkdirpSync(dirname(pythonFile));
    writeFileSync(pythonFile, lines.join('\n'));

    this.project.logger.info(`Project definition file was created at ${pythonFile}`);
  }
}

function renderPythonOptions(indent: number, initOptions?: Record<string, any>): string {
  if (!initOptions || Object.keys(initOptions).length === 0) {
    return ''; // no options
  }

  const lines = [''];

  for (const [name, value] of Object.entries(initOptions)) {
    lines.push(`${toPythonProperty(name)}=${toPythonValue(value)},`);
  }

  return lines.join(`\n${' '.repeat((indent + 1) * 4)}`).concat('\n');
}

function toPythonProperty(prop: string) {
  return decamelize(prop);
}

function toPythonValue(value: any) {
  if (typeof value === 'boolean') {
    return value ? 'True' : 'False';
  } else if (typeof value === 'number') {
    return JSON.stringify(value);
  } else if (typeof value === 'string') {
    return JSON.stringify(value);
  } else if (value === undefined || value === null) {
    return 'None';
  } else {
    return JSON.stringify(value);
  }
}

function toPythonImport(moduleName: string, jsiiType: any) {
  const parts = [moduleName, jsiiType.namespace, jsiiType.name].filter(x => x);
  if (parts.length === 1) {
    return `import ${parts[0]}`;
  } else {
    return `from ${parts.slice(0, -1).join('.')} import ${parts[parts.length - 1]}`;
  }
}