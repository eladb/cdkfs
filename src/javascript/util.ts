import { existsSync, readFileSync } from "fs";
import { basename, dirname, extname, join, sep, resolve } from "path";
import * as semver from "semver";
import { findUp } from "../util";

/**
 * Basic interface for `package.json`.
 */
interface PackageManifest {
  [key: string]: any;

  /**
   * Package name.
   */
  name: string;
  /**
   * Package version.
   */
  version: string;
}

export function renderBundleName(entrypoint: string) {
  const parts = join(entrypoint).split(sep);
  if (parts[0] === "src") {
    parts.shift(); // just remove 'src' if its the first element for ergonomics
  }

  const p = parts.join(sep);
  const dir = dirname(p);
  const base = basename(p, extname(p));
  return join(dir, base);
}

/**
 * Regex for AWS CodeArtifact registry
 */
export const codeArtifactRegex =
  /^https:\/\/(?<registry>(?<domain>[^\.]+)-(?<accountId>\d{12})\.d\.codeartifact\.(?<region>[^\.]+).*\.amazonaws\.com\/.*\/(?<repository>[^\/.]+)\/)/;

/**
 * gets AWS details from the Code Artifact registry URL
 * throws exception if not matching expected pattern
 * @param registryUrl Code Artifact registry URL
 * @returns object containing the (domain, accountId, region, repository)
 */
export function extractCodeArtifactDetails(registryUrl: string) {
  const match = registryUrl.match(codeArtifactRegex);
  if (match?.groups) {
    const { domain, accountId, region, repository, registry } = match.groups;
    return { domain, accountId, region, repository, registry };
  }
  throw new Error("Could not get CodeArtifact details from npm Registry");
}

export function minVersion(version: string): string | undefined {
  if (semver.validRange(version)) {
    return semver.minVersion(version)?.version;
  } else {
    return version;
  }
}

/**
 * Attempt to resolve location of the given `moduleId`.
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModule(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  try {
    return require.resolve(moduleId, options);
  } catch {
    return undefined;
  }
}

/**
 * Attempt to resolve a module's manifest (package.json) path via `require.resolve` lookup.
 *
 * @remarks
 * If the target package has `exports` that differ from the default
 * (i.e, it defines the `exports` field in its manifest) and does not
 * explicitly include an entry for `package.json`, this strategy will fail.
 * See {@link tryResolveManifestPathFromDefaultExport} as an alternative.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModuleManifestPath(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  // cannot just `require('dependency/package.json')` here because
  // `options.paths` may not overlap with this node proc's resolution paths.
  const manifestId = `${moduleId}/package.json`;
  return tryResolveModule(manifestId, options);
}

/**
 * Attempt to resolve a module's manifest (package.json) path by looking for the nearest
 * `package.json` file that is an ancestor to the module's default export location.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveManifestPathFromDefaultExport(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const defaultExportPath = tryResolveModule(moduleId, options);
  if (!defaultExportPath) {
    return undefined;
  }
  const moduleDir = findUp("package.json", defaultExportPath);
  if (!moduleDir) {
    return undefined;
  }
  return join(moduleDir, "package.json");
}

/**
 * Attempt to resolve a module's manifest (package.json) path by checking for its existence under `node_modules` relative to `basePath`.
 *
 * @remarks
 * This strategy can be helpful in the scenario that a module defines
 * custom exports without `package.json` and no default export (i.e, some type definition packages).
 *
 * @param moduleId Module ID to lookup.
 * @param basePath Root path to search from.
 */
export function tryResolveManifestPathFromPath(
  moduleId: string,
  basePath: string
) {
  const base = basePath.includes("node_modules")
    ? basePath
    : join(basePath, "node_modules");
  const filePath = resolve(base, ...moduleId.split("/"), "package.json");
  if (existsSync(filePath)) {
    return filePath;
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) path by searching for it in the optionally provided paths array
 * as well as the current node processes' default resolution paths.
 * @param moduleId Module ID to search for.
 * @param options Search options.
 */
export function tryResolveManifestPathFromSearch(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const searchPaths = [
    ...(options?.paths ?? []),
    ...(require.resolve.paths(moduleId) ?? []),
  ];
  for (const path of searchPaths) {
    const result = tryResolveManifestPathFromPath(moduleId, path);
    // early return on first result.
    if (result) {
      return result;
    }
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) using multiple strategies.
 * @param moduleId Module to resolve manifest path for.
 * @param options Resolution options.
 */
export function tryResolveModuleManifest(
  moduleId: string,
  options?: { paths: string[] }
): PackageManifest | undefined {
  const strategies = [
    tryResolveModuleManifestPath,
    tryResolveManifestPathFromDefaultExport,
    tryResolveManifestPathFromSearch,
  ];
  for (const strategy of strategies) {
    const result = strategy(moduleId, options);
    // early return on first result.
    if (result) {
      try {
        const manifest = JSON.parse(
          readFileSync(result, "utf8")
        ) as PackageManifest;
        // verify name matches target module.
        if (manifest.name === moduleId) {
          return manifest;
        }
      } catch {
        // continue to next strategy.
      }
    }
  }
  return undefined;
}

/**
 * Attempt to resolve the installed version of a given dependency.
 * @param dependencyName Name of dependency.
 * @param options Optional options passed through to `require.resolve`.
 */
export function tryResolveDependencyVersion(
  dependencyName: string,
  options?: { paths: string[] }
): string | undefined {
  const manifest = tryResolveModuleManifest(dependencyName, options);
  if (!manifest) {
    return undefined;
  }
  return manifest?.version;
}

/**
 * Whether the given requestedRange *probably* leads to the installation of a version that matches checkRange
 *
 * We assume that NPM always installs the most recent version of a package that
 * is allowed by the requestedRange.
 */
export function installedVersionProbablyMatches(
  requestedRange: string,
  checkRange: string
): boolean {
  const options = {
    includePrerelease: true,
    loose: true,
  };

  // No questions asked: always true
  if (semver.subset(requestedRange, checkRange, options)) {
    return true;
  }

  // Also no questions asked: always false
  if (!semver.intersects(requestedRange, checkRange, options)) {
    return false;
  }

  // Now we're in tricky territory. We intersect, but aren't a full subset.
  // We are in one of the following 2 situations, which we will tie-break by
  // assuming NPM will install the most recent matching version in 'requested'.
  //
  // requested  | check    | result
  // -----------|----------|-----------
  //   >= 2     |  >= 3    | probably true (chance of FP)
  //   <= 2     |  <= 1    | probably false (change of FN)
  //
  // `semver` doesn't make it easy to distinguish these cases (we can't request
  // the `maxVersion` that satisfies a range). Instead what we do is
  // get the `minVersion` of each range, and if they compare equal we assume
  // we're in the bottom case with `<=` checks, and return `false`.

  return !semver.eq(
    semver.minVersion(requestedRange, options) ?? "1.2.3",
    semver.minVersion(checkRange, options) ?? "1.2.3"
  );
}
