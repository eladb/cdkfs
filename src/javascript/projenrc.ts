import { existsSync, writeFileSync } from 'fs';
import { resolve } from 'path';
import { Component } from '../component';
import { renderJavaScriptOptions } from '../javascript/render-options';
import { Project } from '../project';
import { TypeScriptProject } from '../typescript';

export interface ProjenrcOptions {
  /**
   * The name of the projenrc file.
   * @default ".projenrc.js"
   */
  readonly filename?: string;

  /**
   * Include commented out properties.
   * @default true
   */
  readonly comments?: boolean;
}

/**
 * Sets up a typescript project to use TypeScript for projenrc.
 */
export class Projenrc extends Component {
  private readonly rcfile: string;
  private readonly comments?: boolean;

  constructor(project: Project, options: ProjenrcOptions = {}) {
    super(project);

    this.rcfile = options.filename ?? '.projenrc.js';
    this.comments = options.comments ?? true;

    // this is the task projen executes when running `projen`
    project.addTask(TypeScriptProject.DEFAULT_TASK, { exec: `node ${this.rcfile}` });

    this.generateProjenrc();
  }

  private generateProjenrc() {
    const rcfile = resolve(this.project.outdir, this.rcfile);
    if (existsSync(rcfile)) {
      return; // already exists
    }

    const bootstrap = this.project.newProject;
    if (!bootstrap) {
      return;
    }

    const parts = bootstrap.fqn.split('.');
    const moduleName = parts[0];
    const importName = parts[1];
    const className = parts.slice(1).join('.');

    const js = renderJavaScriptOptions({
      comments: this.comments,
      args: bootstrap.args,
      type: bootstrap.type,
    });

    const lines = new Array<string>();
    lines.push(`const { ${importName} } = require('${moduleName}');`);
    lines.push();
    lines.push(`const project = new ${className}(${js});`);
    lines.push();
    lines.push('project.synth();');

    writeFileSync(rcfile, lines.join('\n'));
    this.project.logger.info(`Project definition file was created at ${rcfile}`);
  }
}