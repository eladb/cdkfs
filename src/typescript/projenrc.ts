import { existsSync, writeFileSync } from 'fs';
import { resolve } from 'path';
import { Component } from '../component';
import { renderJavaScriptOptions } from '../javascript/render-options';
import { TypeScriptProject } from '../typescript';

export interface ProjenrcOptions {
  /**
   * The name of the projenrc file.
   * @default ".projenrc.ts"
   */
  readonly filename?: string;
}

/**
 * Sets up a typescript project to use TypeScript for projenrc.
 */
export class Projenrc extends Component {
  private readonly rcfile: string;

  constructor(project: TypeScriptProject, options: ProjenrcOptions = {}) {
    super(project);

    this.rcfile = options.filename ?? '.projenrc.ts';

    // tell eslint to take .projenrc.ts into account as a dev-dependency
    project.eslint?.allowDevDeps(this.rcfile);
    project.eslint?.addIgnorePattern(`!${this.rcfile}`);
    project.tsconfigEslint?.addInclude(this.rcfile);

    // this is the task projen executes when running `projen` without a
    // specific task (if this task is not defined, projen falls back to
    // running "node .projenrc.js").
    project.addDevDeps('ts-node@^9');
    project.addTask(TypeScriptProject.DEFAULT_TASK, { exec: `ts-node ${this.rcfile}` });

    this.generateProjenrc();
  }

  private generateProjenrc() {
    const rcfile = resolve(this.project.outdir, this.rcfile);
    if (existsSync(rcfile)) {
      return; // already exists
    }

    const bootstrap = this.project.newProject;
    if (!bootstrap) {
      return;
    }

    const parts = bootstrap.fqn.split('.');
    const moduleName = parts[0];
    const importName = parts[1];
    const className = parts.slice(1).join('.');

    const { renderedOptions, imports } = renderJavaScriptOptions({
      args: bootstrap.args,
      type: bootstrap.type,
      comments: bootstrap.comments,
    });

    imports.add(importName);

    const lines = new Array<string>();
    lines.push(`import { ${[...imports].sort().join(', ')} } from '${moduleName}';`);
    lines.push();
    lines.push(`const project = new ${className}(${renderedOptions});`);
    lines.push();
    lines.push('project.synth();');

    writeFileSync(rcfile, lines.join('\n'));
    this.project.logger.info(`Project definition file was created at ${rcfile}`);
  }
}