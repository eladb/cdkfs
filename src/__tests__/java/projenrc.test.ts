import { Pom } from '../../java';
import { Projenrc } from '../../java/projenrc';
import { synthSnapshot, TestProject } from '../util';

test('projenrc.java support', () => {
  // GIVEN
  const project = new TestProject();
  const pom = new Pom(project, {
    groupId: 'my.group.id',
    artifactId: 'hello-world',
    version: '1.2.3',
  });

  // WHEN
  new Projenrc(project, pom);

  // THEN
  expect(synthSnapshot(project)).toMatchSnapshot();
});

test('use a non-test scope', () => {
  // GIVEN
  const project = new TestProject();
  const pom = new Pom(project, {
    groupId: 'my.group.id',
    artifactId: 'hello-world',
    version: '1.2.3',
  });

  // WHEN
  new Projenrc(project, pom, {
    testScope: false,
  });

  // THEN
  expect(synthSnapshot(project)).toMatchSnapshot();
});

test('set the class name', () => {
  // GIVEN
  const project = new TestProject();
  const pom = new Pom(project, {
    groupId: 'my.group.id',
    artifactId: 'hello-world',
    version: '1.2.3',
  });

  // WHEN
  new Projenrc(project, pom, {
    className: 'boom.bam.projenrc',
  });

  // THEN
  expect(synthSnapshot(project)).toMatchSnapshot();
});