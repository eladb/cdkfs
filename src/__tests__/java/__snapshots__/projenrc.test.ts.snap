// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate projenrc in java 1`] = `
"import org.projen.java.JavaProject;
import org.projen.java.JavaProjectOptions;

public class projenrc {
    public static void main(String[] args) {
        JavaProject project = new JavaProject();
        project.synth();
    }
}"
`;

exports[`projenrc.java support 1`] = `
Object {
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

/.gitattributes	linguist-generated
/.projen/      	linguist-generated
/pom.xml       	linguist-generated",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
node_modules/
!/.gitattributes
!/.projen/deps.json
!/.projen/tasks.json
!/pom.xml
",
  ".projen/deps.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "dependencies": Array [
      Object {
        "name": "org.codehaus.mojo/exec-maven-plugin",
        "type": "build",
        "version": "3.0.0",
      },
      Object {
        "name": "com.github.eladb/projen",
        "type": "test",
        "version": "1.0.0",
      },
    ],
  },
  ".projen/tasks.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "tasks": Object {
      "default": Object {
        "description": "Synthesize the project",
        "name": "default",
        "steps": Array [
          Object {
            "exec": "mvn compiler:testCompile --quiet",
          },
          Object {
            "exec": "mvn exec:java --quiet -Dexec.mainClass=projenrc -Dexec.classpathScope=\\"test\\"",
          },
        ],
      },
    },
  },
  "README.md": "# replace this",
  "pom.xml": "<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<project xsi:schemaLocation=\\"http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd\\" xmlns=\\"http://maven.apache.org/POM/4.0.0\\" xmlns:xsi=\\"http://www.w3.org/2001/XMLSchema-instance\\">
    <modelVersion>4.0.0</modelVersion>
    <groupId>my.group.id</groupId>
    <artifactId>hello-world</artifactId>
    <version>1.2.3</version>
    <packaging>jar</packaging>
    <name>my-project</name>
    <build>
        <plugins>
            <plugin>
                <groupId>org.codehaus.mojo</groupId>
                <artifactId>exec-maven-plugin</artifactId>
                <version>3.0.0</version>
            </plugin>
        </plugins>
    </build>
    <dependencies>
        <dependency>
            <groupId>com.github.eladb</groupId>
            <artifactId>projen</artifactId>
            <version>1.0.0</version>
            <scope>test</scope>
        </dependency>
    </dependencies>
</project>
<!-- ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\". -->",
}
`;

exports[`set the class name 1`] = `
Object {
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

/.gitattributes	linguist-generated
/.projen/      	linguist-generated
/pom.xml       	linguist-generated",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
node_modules/
!/.gitattributes
!/.projen/deps.json
!/.projen/tasks.json
!/pom.xml
",
  ".projen/deps.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "dependencies": Array [
      Object {
        "name": "org.codehaus.mojo/exec-maven-plugin",
        "type": "build",
        "version": "3.0.0",
      },
      Object {
        "name": "com.github.eladb/projen",
        "type": "test",
        "version": "7.7.7",
      },
    ],
  },
  ".projen/tasks.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "tasks": Object {
      "default": Object {
        "description": "Synthesize the project",
        "name": "default",
        "steps": Array [
          Object {
            "exec": "mvn compiler:testCompile --quiet",
          },
          Object {
            "exec": "mvn exec:java --quiet -Dexec.mainClass=boom.bam.projenrc -Dexec.classpathScope=\\"test\\"",
          },
        ],
      },
    },
  },
  "README.md": "# replace this",
  "pom.xml": "<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<project xsi:schemaLocation=\\"http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd\\" xmlns=\\"http://maven.apache.org/POM/4.0.0\\" xmlns:xsi=\\"http://www.w3.org/2001/XMLSchema-instance\\">
    <modelVersion>4.0.0</modelVersion>
    <groupId>my.group.id</groupId>
    <artifactId>hello-world</artifactId>
    <version>1.2.3</version>
    <packaging>jar</packaging>
    <name>my-project</name>
    <build>
        <plugins>
            <plugin>
                <groupId>org.codehaus.mojo</groupId>
                <artifactId>exec-maven-plugin</artifactId>
                <version>3.0.0</version>
            </plugin>
        </plugins>
    </build>
    <dependencies>
        <dependency>
            <groupId>com.github.eladb</groupId>
            <artifactId>projen</artifactId>
            <version>7.7.7</version>
            <scope>test</scope>
        </dependency>
    </dependencies>
</project>
<!-- ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\". -->",
}
`;

exports[`use a non-test scope 1`] = `
Object {
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

/.gitattributes	linguist-generated
/.projen/      	linguist-generated
/pom.xml       	linguist-generated",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
node_modules/
!/.gitattributes
!/.projen/deps.json
!/.projen/tasks.json
!/pom.xml
",
  ".projen/deps.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "dependencies": Array [
      Object {
        "name": "org.codehaus.mojo/exec-maven-plugin",
        "type": "build",
        "version": "3.0.0",
      },
      Object {
        "name": "com.github.eladb/projen",
        "type": "runtime",
        "version": "7.7.7",
      },
    ],
  },
  ".projen/tasks.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "tasks": Object {
      "default": Object {
        "description": "Synthesize the project",
        "name": "default",
        "steps": Array [
          Object {
            "exec": "mvn compiler:compile --quiet",
          },
          Object {
            "exec": "mvn exec:java --quiet -Dexec.mainClass=projenrc",
          },
        ],
      },
    },
  },
  "README.md": "# replace this",
  "pom.xml": "<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<project xsi:schemaLocation=\\"http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd\\" xmlns=\\"http://maven.apache.org/POM/4.0.0\\" xmlns:xsi=\\"http://www.w3.org/2001/XMLSchema-instance\\">
    <modelVersion>4.0.0</modelVersion>
    <groupId>my.group.id</groupId>
    <artifactId>hello-world</artifactId>
    <version>1.2.3</version>
    <packaging>jar</packaging>
    <name>my-project</name>
    <build>
        <plugins>
            <plugin>
                <groupId>org.codehaus.mojo</groupId>
                <artifactId>exec-maven-plugin</artifactId>
                <version>3.0.0</version>
            </plugin>
        </plugins>
    </build>
    <dependencies>
        <dependency>
            <groupId>com.github.eladb</groupId>
            <artifactId>projen</artifactId>
            <version>7.7.7</version>
        </dependency>
    </dependencies>
</project>
<!-- ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\". -->",
}
`;
